presto-main/src/main/java/com/facebook/presto/execution/QueueEntry.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.concurrent.SetThreadName;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class QueueEntry
+{
+    private final QueryExecution queryExecution;
+    private final List<QueryQueue> nextQueues;
+    private final ListenableFuture<?> listenableFuture;
+
+    public QueueEntry(QueryExecution queryExecution, List<QueryQueue> nextQueues)
+    {
+        this.queryExecution = checkNotNull(queryExecution, "queryExecution is null");
+        this.nextQueues = ImmutableList.copyOf(checkNotNull(nextQueues, "nextQueues is null"));
+        SettableFuture<?> settableFuture = SettableFuture.create();
+        SqlQueryManager.addCompletionCallback(queryExecution, () -> settableFuture.set(null));
+        listenableFuture = settableFuture;
```

Comment:
it looks like we've lost the ability to cancel an item in the queue by dereferencing it? did we do that on purpose?